shm/test_ucp_tag_stats.offload_sw_rndv/0 failed

```
09:19:06 [ RUN      ] shm/test_ucp_tag_stats.offload_sw_rndv/0
09:19:06 /scrap/jenkins/workspace/hpc-ucx-pr/label/clx-ppc-04.mtl.labs.mlnx/worker/2/contrib/../test/gtest/ucp/test_ucp_tag_xfer.cc:930: Failure
09:19:06 Value of: cnt
09:19:06   Actual: 0
09:19:06 Expected: 1ul
09:19:06 Which is: 1
09:19:06 
[  FAILED  ] shm/test_ucp_tag_stats.offload_sw_rndv/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (342 ms)
```

full log: http://hpc-master.lab.mtl.com:8080/view/HPC/job/hpc-ucx-pr/5632/label=clx-ppc-04.mtl.labs.mlnx,worker=2/consoleFull#-94122935044550c5d-247a-4f9d-bad5-5b93d2b6e2ac
failed in not related PR https://github.com/openucx/ucx/pull/2087

failed on x86 as well:
http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5728/label=clx-ppc-04.mtl.labs.mlnx,worker=2/console
```
12:27:29 [       OK ] shm_ib/test_ucp_tag_stats.sync_unexpected/0 (395 ms)
12:27:29 [ RUN      ] shm_ib/test_ucp_tag_stats.offload_sw_rndv/0
12:27:30 /scrap/jenkins/workspace/hpc-ucx-pr-2/label/clx-ppc-04.mtl.labs.mlnx/worker/2/contrib/../test/gtest/ucp/test_ucp_tag_xfer.cc:930: Failure
12:27:30 Value of: cnt
12:27:30   Actual: 0
12:27:30 Expected: 1ul
12:27:30 Which is: 1
12:27:30 
[  FAILED  ] shm_ib/test_ucp_tag_stats.offload_sw_rndv/0, where GetParam() = shm,ib (452 ms)
12:27:30 [----------] 2 tests from shm_ib/test_ucp_tag_stats (848 ms total)
...
12:29:13 
make: *** [test] Error 1
12:29:13 make: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr-2/label/clx-ppc-04.mtl.labs.mlnx/worker/2/build-test/test/gtest'
```

--- a/test/gtest/ucp/test_ucp_tag_offload.cc
+++ b/test/gtest/ucp/test_ucp_tag_offload.cc
@@ -21,7 +21,6 @@ public:
     {
         m_env.push_back(new ucs::scoped_setenv("UCX_RC_TM_ENABLE", "y"));
         modify_config("TM_OFFLOAD", "y");
-        modify_config("TM_THRESH" , "1024");
 
         test_ucp_tag::init();
         if (!(sender().ep()->flags & UCP_EP_FLAG_TAG_OFFLOAD_ENABLED)) {
@@ -110,7 +109,7 @@ UCS_TEST_P(test_ucp_tag_offload, post_after_comp)
 
     send_b(&small_val, sizeof(small_val), DATATYPE, 0x11);
     wait(req);
-    request_release(req);
+    request_free(req);
 
     req = recv_nb_and_check(&recvbuf, recvbuf.size(), DATATYPE, tag,
                             UCP_TAG_MASK_FULL);
@@ -174,8 +173,7 @@ UCS_TEST_P(test_ucp_tag_offload, post_dif_buckets)
     EXPECT_EQ(2u, receiver().worker()->tm.expected.sw_all_count);
 
     for (std::vector<request*>::const_iterator iter = reqs.begin();
-         iter != reqs.end(); ++iter)
-    {
+         iter != reqs.end(); ++iter) {
         req_cancel(receiver(), *iter);
     }
 }
@@ -281,3 +279,179 @@ UCS_TEST_P(test_ucp_tag_offload_multi, recv_from_multi)
 
 UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_tag_offload_multi, all_rcdc,
                               "\\rc,\\dc,\\ud,rc_x,dc_x")
+
+
+#if ENABLE_STATS
+
+class test_ucp_tag_offload_stats : public test_ucp_tag_offload_multi {
+public:
+
+    void init()
+    {
+        stats_activate();
+        test_ucp_tag_offload::init(); // No need for multi::init()
+    }
+
+    void cleanup()
+    {
+        test_ucp_tag_offload::cleanup();
+        stats_restore();
+    }
+
+    request* recv_nb_exp(void *buffer, size_t count, ucp_datatype_t dt,
+                         ucp_tag_t tag, ucp_tag_t tag_mask)
+    {
+        request *req1 = recv_nb_and_check(buffer, count, DATATYPE, tag,
+                                          UCP_TAG_MASK_FULL);
+
+        // Post and cancel another receive to make sure the first one was offloaded
+        request *req2 = recv_nb_and_check(buffer, count, DATATYPE, tag,
+                                          UCP_TAG_MASK_FULL);
+        req_cancel(receiver(), req2);
+
+        return req1;
+    }
+
+    ucs_stats_node_t* worker_offload_stats(entity &e)
+    {
+        return e.worker()->tm_offload_stats;
+    }
+
+    void validate_offload_counter(uint64_t rx_cntr, uint64_t val)
+    {
+        uint64_t cnt;
+        cnt = UCS_STATS_GET_COUNTER(worker_offload_stats(receiver()), rx_cntr);
+        EXPECT_EQ(val, cnt);
+    }
+
+    void wait_counter(ucs_stats_node_t *stats, uint64_t cntr,
+                      double timeout = UCP_TEST_TIMEOUT_IN_SEC)
+    {
+        ucs_time_t deadline = ucs_get_time() + ucs_time_from_sec(timeout);
+        uint64_t v;
+
+        do {
+            short_progress_loop();
+            v = UCS_STATS_GET_COUNTER(stats, cntr);
+        } while ((ucs_get_time() < deadline) && !v);
+
+        EXPECT_EQ(1ul, v);
+    }
+
+    void test_send_recv(size_t count, bool send_iov, uint64_t cntr)
+    {
+        ucp_tag_t tag = 0x11;
+
+        std::vector<char> sbuf(count, 0);
+        std::vector<char> rbuf(count, 0);
+        request *req = recv_nb_exp(rbuf.data(), rbuf.size(), DATATYPE, tag,
+                                   UCP_TAG_MASK_FULL);
+
+        if (send_iov) {
+            ucp::data_type_desc_t dt_desc(DATATYPE_IOV, sbuf.data(),
+                                          sbuf.size(), 1);
+            send_b(dt_desc.buf(), dt_desc.count(), dt_desc.dt(), tag);
+        } else {
+            send_b(sbuf.data(), sbuf.size(), DATATYPE, tag);
+        }
+        wait(req);
+        request_free(req);
+
+        validate_offload_counter(cntr, 1ul);
+    }
+};
+
+UCS_TEST_P(test_ucp_tag_offload_stats, post, "TM_THRESH=1")
+{
+    uint64_t dummy;
+    uint64_t tag = 0x11;
+
+    activate_offload(sender());
+
+    request *rreq = recv_nb(&dummy, sizeof(dummy), DATATYPE, tag,
+                            UCP_TAG_MASK_FULL);
+
+    wait_counter(worker_offload_stats(receiver()),
+                 UCP_WORKER_STAT_TAG_OFFLOAD_POSTED);
+
+    req_cancel(receiver(), rreq);
+
+    wait_counter(worker_offload_stats(receiver()),
+                 UCP_WORKER_STAT_TAG_OFFLOAD_CANCELED);
+}
+
+UCS_TEST_P(test_ucp_tag_offload_stats, block, "TM_THRESH=1")
+{
+    uint64_t tag = 0x11;
+    std::vector<char> buf(64, 0);
+
+    activate_offload(sender());
+
+    // Check BLOCK_NON_CONTIG
+    ucp::data_type_desc_t dt_desc(DATATYPE_IOV, buf.data(), buf.size(), 1);
+    request *rreq = recv_nb_and_check(dt_desc.buf(), dt_desc.count(),
+                                      dt_desc.dt(), tag, UCP_TAG_MASK_FULL);
+
+    wait_counter(worker_offload_stats(receiver()),
+                 UCP_WORKER_STAT_TAG_OFFLOAD_BLOCK_NON_CONTIG);
+
+    req_cancel(receiver(), rreq);
+
+    // Check BLOCK_WILDCARD
+    rreq = recv_nb_and_check(buf.data(), buf.size(), DATATYPE, tag, 0);
+
+    wait_counter(worker_offload_stats(receiver()),
+                 UCP_WORKER_STAT_TAG_OFFLOAD_BLOCK_WILDCARD);
+
+    req_cancel(receiver(), rreq);
+
+    // Check BLOCK_TAG_EXCEED
+    std::vector<request*> reqs;
+    uint64_t cnt;
+    unsigned limit = 1000; // Just a big value to avoid test hang
+    do {
+        rreq = recv_nb_and_check(buf.data(), buf.size(), DATATYPE, tag,
+                                 UCP_TAG_MASK_FULL);
+        cnt  = UCS_STATS_GET_COUNTER(worker_offload_stats(receiver()),
+                                    UCP_WORKER_STAT_TAG_OFFLOAD_BLOCK_TAG_EXCEED);
+        reqs.push_back(rreq);
+    } while (!cnt && (--limit > 0));
+
+    validate_offload_counter(UCP_WORKER_STAT_TAG_OFFLOAD_BLOCK_TAG_EXCEED , 1ul);
+
+    for (std::vector<request*>::const_iterator iter = reqs.begin();
+         iter != reqs.end(); ++iter) {
+        req_cancel(receiver(), *iter);
+    }
+}
+
+UCS_TEST_P(test_ucp_tag_offload_stats, eager, "RNDV_THRESH=1000", "TM_THRESH=64")
+{
+    size_t size = 512; // Size smaller than RNDV, but bigger than TM thresh
+
+    // Offload is not activated, so the first message should arrive unexpectedly
+    test_send_recv(size, false, UCP_WORKER_STAT_TAG_OFFLOAD_RX_UNEXP_EGR);
+    test_send_recv(size, false, UCP_WORKER_STAT_TAG_OFFLOAD_MATCHED);
+}
+
+UCS_TEST_P(test_ucp_tag_offload_stats, rndv, "RNDV_THRESH=1000")
+{
+    size_t size = 2048; // Size bigger than RNDV thresh
+
+    // Offload is not activated, so the first message should arrive unexpectedly
+    test_send_recv(size, false, UCP_WORKER_STAT_TAG_OFFLOAD_RX_UNEXP_RNDV);
+    test_send_recv(size, false, UCP_WORKER_STAT_TAG_OFFLOAD_MATCHED);
+}
+
+UCS_TEST_P(test_ucp_tag_offload_stats, sw_rndv, "RNDV_THRESH=1000")
+{
+    size_t size = 2048; // Size bigger than RNDV thresh
+
+    // Offload is not activated, so the first message should arrive unexpectedly
+    test_send_recv(size, true, UCP_WORKER_STAT_TAG_OFFLOAD_RX_UNEXP_SW_RNDV);
+    test_send_recv(size, true, UCP_WORKER_STAT_TAG_OFFLOAD_MATCHED_SW_RNDV);
+}
+
+UCP_INSTANTIATE_TEST_CASE(test_ucp_tag_offload_stats)
+
+#endif

--- a/test/gtest/ucp/test_ucp_tag_xfer.cc
+++ b/test/gtest/ucp/test_ucp_tag_xfer.cc
@@ -913,30 +913,11 @@ public:
         return e.worker()->stats;
     }
 
-    ucs_stats_node_t* worker_offload_stats(entity &e) {
-        return e.worker()->tm_offload_stats;
-    }
-
-    void skip_no_tag_offload() {
-        if (!(sender().ep()->flags & UCP_EP_FLAG_TAG_OFFLOAD_ENABLED)) {
-            UCS_TEST_SKIP_R("no tag offload");
-        }
-    }
-
     void validate_counters(uint64_t tx_cntr, uint64_t rx_cntr) {
         uint64_t cnt;
         cnt = UCS_STATS_GET_COUNTER(ep_stats(sender()), tx_cntr);
         EXPECT_EQ(1ul, cnt);
         cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()), rx_cntr);
-        EXPECT_EQ(1ul, cnt);
-    }
-
-    void validate_offload_counters(uint64_t rx_cntr, uint64_t posted_cnt) {
-        uint64_t cnt;
-        cnt = UCS_STATS_GET_COUNTER(worker_offload_stats(receiver()),
-                                    UCP_WORKER_STAT_TAG_OFFLOAD_POSTED);
-        EXPECT_EQ(posted_cnt, cnt);
-        cnt = UCS_STATS_GET_COUNTER(worker_offload_stats(receiver()), rx_cntr);
         EXPECT_EQ(1ul, cnt);
     }
 
@@ -1016,75 +997,6 @@ UCS_TEST_P(test_ucp_tag_stats, rndv_unexpected, "RNDV_THRESH=1000",
     test_run_xfer(true, true, false, false, false);
     validate_counters(UCP_EP_STAT_TAG_TX_RNDV,
                       UCP_WORKER_STAT_TAG_RX_RNDV_UNEXP);
-}
-
-UCS_TEST_P(test_ucp_tag_stats, offload_rndv, "RNDV_THRESH=1000",
-                                             "TM_THRESH=1",
-                                             "TM_OFFLOAD=y") {
-    skip_no_tag_offload();
-
-    test_run_xfer(true, true, true, false, false);
-
-    // First tag offload packet should be arrived unexpectedly in any case
-    validate_offload_counters(UCP_WORKER_STAT_TAG_OFFLOAD_RX_UNEXP_RNDV, 0ul);
-
-    test_run_xfer(true, true, true, false, false);
-
-    // Now the packet should arrive expectedly, since previous packet activated
-    // offload interface
-    validate_offload_counters(UCP_WORKER_STAT_TAG_OFFLOAD_MATCHED, 1ul);
-}
-
-UCS_TEST_P(test_ucp_tag_stats, offload_sw_rndv, "RNDV_THRESH=1000",
-                                                "TM_THRESH=1",
-                                                "TM_OFFLOAD=y") {
-    skip_no_tag_offload();
-
-    test_run_xfer(false, true, true, false, false);
-
-    validate_offload_counters(UCP_WORKER_STAT_TAG_OFFLOAD_RX_UNEXP_SW_RNDV, 0ul);
-
-    test_run_xfer(false, true, true, false, false);
-
-    validate_offload_counters(UCP_WORKER_STAT_TAG_OFFLOAD_MATCHED_SW_RNDV, 1ul);
-}
-
-UCS_TEST_P(test_ucp_tag_stats, offload_exp_recv_generic, "RNDV_THRESH=1000",
-                                                         "TM_THRESH=1",
-                                                         "TM_OFFLOAD=y") {
-    skip_no_tag_offload();
-
-    // Send unexpected msg to activate offload interface
-    test_run_xfer(false, true, false, false, false);
-
-    test_run_xfer(false, false, true, false, false);
-
-    validate_offload_counters(UCP_WORKER_STAT_TAG_OFFLOAD_BLOCK_NON_CONTIG, 0ul);
-}
-
-UCS_TEST_P(test_ucp_tag_stats, offload_post, "TM_OFFLOAD=y", "TM_THRESH=1") {
-
-    skip_no_tag_offload();
-
-    uint64_t dummy;
-    uint64_t tag  = 0x11;
-    uint64_t mask = 0xffff;
-
-    // Send unexpected msg to activate offload interface
-    test_run_xfer(false, true, false, false, false);
-
-    // Check general flow
-    request *rreq = recv_nb(&dummy, sizeof(dummy), DATATYPE, tag, mask);
-
-    wait_counter(worker_offload_stats(receiver()),
-                 UCP_WORKER_STAT_TAG_OFFLOAD_POSTED);
-
-    ucp_request_cancel(receiver().worker(), rreq);
-
-    wait_counter(worker_offload_stats(receiver()),
-                 UCP_WORKER_STAT_TAG_OFFLOAD_CANCELED);
-
-    request_release(rreq);
 }
 
 UCP_INSTANTIATE_TEST_CASE(test_ucp_tag_stats)

--- a/test/gtest/ucp/test_ucp_tag_xfer.cc
+++ b/test/gtest/ucp/test_ucp_tag_xfer.cc
@@ -921,19 +921,6 @@ public:
         EXPECT_EQ(1ul, cnt);
     }
 
-    void wait_counter(ucs_stats_node_t *stats, uint64_t cntr,
-                      double timeout = UCP_TEST_TIMEOUT_IN_SEC) {
-        ucs_time_t deadline = ucs_get_time() + ucs_time_from_sec(timeout);
-        uint64_t v;
-
-        do {
-            short_progress_loop();
-            v = UCS_STATS_GET_COUNTER(stats, cntr);
-        } while ((ucs_get_time() < deadline) && !v);
-
-        EXPECT_EQ(1ul, v);
-    }
-
 };
 
 
@@ -943,10 +930,10 @@ UCS_TEST_P(test_ucp_tag_stats, eager_expected, "RNDV_THRESH=1248576",
     validate_counters(UCP_EP_STAT_TAG_TX_EAGER,
                       UCP_WORKER_STAT_TAG_RX_EAGER_MSG);
 
-     uint64_t cnt;
-     cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
-                                 UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
-     EXPECT_EQ(cnt, 0ul);
+    uint64_t cnt;
+    cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
+                                UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
+    EXPECT_EQ(cnt, 0ul);
 }
 
 UCS_TEST_P(test_ucp_tag_stats, eager_unexpected, "RNDV_THRESH=1248576",
@@ -954,10 +941,10 @@ UCS_TEST_P(test_ucp_tag_stats, eager_unexpected, "RNDV_THRESH=1248576",
     test_run_xfer(true, true, false, false, false);
     validate_counters(UCP_EP_STAT_TAG_TX_EAGER,
                       UCP_WORKER_STAT_TAG_RX_EAGER_MSG);
-     uint64_t cnt;
-     cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
-                                 UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
-     EXPECT_GT(cnt, 0ul);
+    uint64_t cnt;
+    cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
+                                UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
+    EXPECT_GT(cnt, 0ul);
 }
 
 UCS_TEST_P(test_ucp_tag_stats, sync_expected, "RNDV_THRESH=1248576",
@@ -967,10 +954,10 @@ UCS_TEST_P(test_ucp_tag_stats, sync_expected, "RNDV_THRESH=1248576",
     validate_counters(UCP_EP_STAT_TAG_TX_EAGER_SYNC,
                       UCP_WORKER_STAT_TAG_RX_EAGER_SYNC_MSG);
 
-     uint64_t cnt;
-     cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
+    uint64_t cnt;
+    cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
                                  UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
-     EXPECT_EQ(cnt, 0ul);
+    EXPECT_EQ(cnt, 0ul);
 }
 
 UCS_TEST_P(test_ucp_tag_stats, sync_unexpected, "RNDV_THRESH=1248576",
@@ -979,10 +966,10 @@ UCS_TEST_P(test_ucp_tag_stats, sync_unexpected, "RNDV_THRESH=1248576",
     test_run_xfer(true, true, false, true, false);
     validate_counters(UCP_EP_STAT_TAG_TX_EAGER_SYNC,
                       UCP_WORKER_STAT_TAG_RX_EAGER_SYNC_MSG);
-     uint64_t cnt;
-     cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
-                                 UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
-     EXPECT_GT(cnt, 0ul);
+    uint64_t cnt;
+    cnt = UCS_STATS_GET_COUNTER(worker_stats(receiver()),
+                                UCP_WORKER_STAT_TAG_RX_EAGER_CHUNK_UNEXP);
+    EXPECT_GT(cnt, 0ul);
 }
 
 UCS_TEST_P(test_ucp_tag_stats, rndv_expected, "RNDV_THRESH=1000",

